internal/components/meshgateway/meshgateway_test.go

```diff
@@ -0,0 +1,137 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package meshgateway_test
+
+import (
+	_ "embed"
+	"fmt"
+	"os"
+	"reflect"
+	"testing"
+
+	"emperror.dev/errors"
+	"emperror.dev/errors/utils/keyval"
+	"github.com/homeport/dyff/pkg/dyff"
+	"k8s.io/client-go/kubernetes/fake"
+	"sigs.k8s.io/yaml"
+
+	"github.com/banzaicloud/istio-operator/v2/api/v1alpha1"
+	"github.com/banzaicloud/istio-operator/v2/internal/assets"
+	"github.com/banzaicloud/istio-operator/v2/internal/components/meshgateway"
+	"github.com/banzaicloud/istio-operator/v2/internal/util"
+	"github.com/banzaicloud/operator-tools/pkg/helm/templatereconciler"
+	"github.com/banzaicloud/operator-tools/pkg/reconciler"
+)
+
+//go:embed testdata/mgw-test-cr.yaml
+var mgwTestCR []byte
+
+//go:embed testdata/mgw-expected-values.yaml
+var mgwExpectedValues []byte
+
+//go:embed testdata/mgw-expected-resource-dump.yaml
+var mgwExpectedResourceDump []byte
+
+func TestMGWResourceDump(t *testing.T) {
+	t.Parallel()
+
+	var mgw *v1alpha1.MeshGateway
+	if err := yaml.Unmarshal(mgwTestCR, &mgw); err != nil {
+		t.Fatal(err)
+	}
+
+	reconciler := meshgateway.NewChartReconciler(
+		templatereconciler.NewHelmReconciler(nil, nil, nil, fake.NewSimpleClientset().Discovery(), []reconciler.NativeReconcilerOpt{
+			reconciler.NativeReconcilerSetControllerRef(),
+		}),
+	)
+
+	dd, err := reconciler.GetManifest(mgw)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	report, err := util.CompareYAMLs(mgwExpectedResourceDump, dd)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(report.Diffs) > 0 {
+		if err := (&dyff.HumanReport{
+			Report:       report,
+			OmitHeader:   false,
+			NoTableStyle: true,
+		}).WriteReport(os.Stdout); err != nil {
+			t.Fatal(err)
+		}
+
+		t.Fatal(errors.NewPlain("generated resource dump not equals with expected"))
+	}
+}
+
+func TestMGWTemplateTransform(t *testing.T) {
+	t.Parallel()
+
+	var mgw *v1alpha1.MeshGateway
+	if err := yaml.Unmarshal(mgwTestCR, &mgw); err != nil {
+		t.Fatal(err)
+	}
+
+	obj := &v1alpha1.MeshGatewayWithProperties{
+		MeshGateway: mgw,
+		Properties: v1alpha1.MeshGatewayProperties{
+			Revision:              "cp-revision-1",
+			EnablePrometheusMerge: false,
+			InjectionTemplate:     "gateway",
+		},
+	}
+
+	values, err := util.TransformICPToStriMapWithTemplate(obj, assets.MeshGateway, "values.yaml.tpl")
+	if err != nil {
+		kv := keyval.ToMap(errors.GetDetails(err))
+		if t, ok := kv["template"]; ok {
+			fmt.Printf("%s\n", t.(string))
+		}
+		t.Fatal(err)
+	}
+
+	var expectedValues map[string]interface{}
+	if err := yaml.Unmarshal(mgwExpectedValues, &expectedValues); err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(values, expectedValues) {
+		valuesYaml, err := yaml.Marshal(values)
+		if err != nil {
+			t.Fatal(err)
+		}
```

Comment:
Is the yaml.Unmarshal + reflect.DeepEqual + yaml.Marshal needed here? Can't you just call the `if len(report.Diffs) > 0 {` after `util.CompareYAMLs(mgwExpectedValues, valuesYaml)`? Similarly how it's done in `TestMGWResourceDump`.